dev/Interop/StoragePickers/PickerCommon.cpp

```diff
@@ -0,0 +1,213 @@
+    {
+        winrt::hstring result;
+        bool first = true;
+        for (const auto& ext : extensions)
+        {
+            if (first)
+            {
+                result = FormatExtensionWithWildcard(ext);
+                first = false;
+            }
+            else
+            {
+                result = result + L";" + FormatExtensionWithWildcard(ext);
+            }
+        }
+        return result;
+    }
+}
+
+
+namespace PickerCommon {
+
+    using namespace winrt;
+
+    bool IsHStringNullOrEmpty(winrt::hstring value)
+    {
+        return value.empty();
+    }
+
+    winrt::hstring GetPathFromShellItem(winrt::com_ptr<IShellItem> shellItem)
+    {
+        wil::unique_cotaskmem_string filePath;
+        check_hresult(shellItem->GetDisplayName(SIGDN_FILESYSPATH, filePath.put()));
+        return winrt::hstring{ filePath.get() };
+    }
+
+    std::vector<COMDLG_FILTERSPEC> CaptureFilterSpec(std::vector<winrt::hstring>& buffer, winrt::Windows::Foundation::Collections::IVectorView<winrt::hstring> filters)
+    {
+        std::vector<COMDLG_FILTERSPEC> result(filters.Size());
+        buffer.clear();
+        buffer.reserve(filters.Size() * (size_t)2);
+        for (auto filter : filters)
+        {
+            auto ext = FormatExtensionWithWildcard(filter);
+            buffer.push_back(filter);
+            buffer.push_back(ext);
+        }
+        for (size_t i = 0; i < filters.Size(); i++)
+        {
+            result.at(i) = { buffer.at(i * 2).c_str(), buffer.at(i * 2 + 1).c_str() };
+        }
+
+        if (result.size() == 0)
+        {
+            result.push_back({ L"All Files", L"*.*" });
+        }
+        return result;
+    }
+
+    std::vector<COMDLG_FILTERSPEC> CaptureFilterSpec(std::vector<winrt::hstring>& buffer, winrt::Windows::Foundation::Collections::IMapView<winrt::hstring, winrt::Windows::Foundation::Collections::IVector<hstring>> filters)
+    {
+        std::vector<COMDLG_FILTERSPEC> result(filters.Size());
+        buffer.clear();
+        buffer.reserve(filters.Size() * 2ull);
+
+        for (const auto& filter : filters)
+        {
+            buffer.push_back(filter.Key());
+            auto extensionList = JoinExtensions(filter.Value().GetView());
+            buffer.push_back(extensionList);
+        }
+        for (size_t i = 0; i < filters.Size(); i++)
+        {
+            result.at(i) = { buffer.at(i * 2).c_str(), buffer.at(i * 2 + 1).c_str() };
+        }
+
+        if (result.empty())
+        {
+            result.push_back({ L"All Files", L"*.*" });
+        }
+        return result;
+    }
+
+    void PickerParameters::ConfigureDialog(winrt::com_ptr<IFileDialog> dialog)
+    {
+        if (!IsHStringNullOrEmpty(CommitButtonText))
+        {
+            check_hresult(dialog->SetOkButtonLabel(CommitButtonText.c_str()));
+        }
+
+        if (!IsHStringNullOrEmpty(SettingsIdentifierId))
+        {
+            auto guid = HashHStringToGuid(SettingsIdentifierId);
```

Comment:
>SettingsIdentifierId [](http://example.com/codeflow?start=42&length=20)

Is SettingsIdentifierId just the Id passed by the app, or is it combined in some way with an app identify (such as package name or exe name if unpackaged)? If it isn't combined with an app identifier, we might want to change that, so you don't get two apps both specifying something simple like "SaveDialog" and slamming each other's settings. Or do packaged apps have this info by default get stored in virtualized registry, such that this might only be an issue for unpackaged apps (and apps which disable registry virtualization)?

(This can be logged as an issue for later follow-up. No need to block initial checkin.)

Reply:
Thank you for your suggestion! A task was created for later follow-up,